src/Types.elm

```diff
@@ -2,18 +2,73 @@ module Types exposing (..)
 
 type Tree a = Node a (List (Tree a))
 
-type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole | TokVar String
-             | TokConstInt Int | TokConstBool Bool | TokInvalid | TokBackSlash | TokArrow | TokEnd
+type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole |
+              TokVar String | TokHasType | TokConstInt Int | TokConstBool Bool |
+              TokInvalid | TokBackSlash | TokArrow | TokEnd | TokTypeName String
 
 type TokTSC = TTSCInt TokTSCInt | TTSCBool TokTSCBool
 type TokTSCInt = TokPlus | TokMinus | TokTimes
 type TokTSCBool = TokEq | TokAnd | TokOr
 
 type Const = CBool Bool | CInt Int
 type Term = CTerm Const | VTerm String | Plus Term Term | Minus Term Term | Times Term Term
-            | Eq Term Term | And Term Term | Or Term Term | Lam Term Term | App Term Term
+            | Eq Term Term | And Term Term | Or Term Term | Lam String Term | App Term Term
             | MissingInt | MissingBool | Missing | EmptyTree
 
+-- V(alue)Type is a type that a TreeAssembly term can evaluate to
+{-
+  TAny may be used when the type can be either int or bool (discuss)
+  TNone is used as a default value for types we haven't determined yet (similar to EmptyTree)
+-}
+type VType = TBool | TInt | TFun VType VType | TAny | TNone
+
 type alias Var =
   { name: String
-  , term: Term }
+  , term: Term
+  , vtype: VType}
+
+{-
+  Returns the type signature of basic language constructs (if it's not relevant, return TNone)
+-}
+getTypeSignature : Term -> VType
+getTypeSignature t =
+  case t of
+    CTerm (CBool b) -> TBool
+    CTerm (CInt i) -> TInt
```

Comment:
These types seem close, but a bit off.  I'm assuming that the `TFun` type is intended as a function from the first type to the second type.  Given this, `Plus` should behave like `TFun TInt (TFun TInt TInt)`, since this will allow for multiple application through currying.  The same goes for the rest of the binary operators.

Reply:
Good catch! I fixed this issue in one of my new commits.